File: skellig-test-step-processing/src/main/java/org/skellig/teststep/processing/model/ValidationDetails.java
package org.skellig.teststep.processing.model;

import java.util.Optional;


public class ValidationDetails {

    private String testStepId;
    private String convertTo;
    private ExpectedResult expectedResult;

    protected ValidationDetails(String testStepId, String convertTo, ExpectedResult expectedResult) {
        this.testStepId = testStepId;
        this.convertTo = convertTo;
        this.expectedResult = expectedResult;
        this.expectedResult.initializeParents();
    }

    public Optional<String> getTestStepId() {
        return Optional.ofNullable(testStepId);
    }

    public Optional<String> getConvertTo() {
        return Optional.ofNullable(convertTo);
    }

    public ExpectedResult getExpectedResult() {
        return expectedResult;
    }

    public static class Builder {
        private String testStepId;
        private String convertTo;
        private ExpectedResult expectedResult;

        public Builder withTestStepId(String testStepId) {
            this.testStepId = testStepId;
            return this;
        }

        public Builder withConvertTo(String convertTo) {
            this.convertTo = convertTo;
            return this;
        }

        public Builder withExpectedResult(ExpectedResult expectedResult) {
            this.expectedResult = expectedResult;
            return this;
        }

        public ValidationDetails build() {
            return new ValidationDetails(testStepId, convertTo, expectedResult);
        }
    }
}

File: skellig-test-step-processing/src/main/kotlin/org/skellig/teststep/processing/model/ValidationDetails.kt
package org.skellig.teststep.processing.model

/**
 * Contains the rules of validating an actual result from a processed test step.
 * These rules are stored in `expectedResult` as a tree structure.
 *
 * @param testStepId an optional property defining which test step result to validate from
 *
 * @see ExpectedResult
 */
data class ValidationDetails(
    val testStepId: String? = null,
    val expectedResult: ExpectedResult
) {

    init {
        expectedResult.initializeParents()
    }

    data class Builder(
        var testStepId: String? = null,
        var convertTo: String? = null,
        var expectedResult: ExpectedResult? = null
    ) {

        fun withTestStepId(testStepId: String?) = apply { this.testStepId = testStepId }

        fun withExpectedResult(expectedResult: ExpectedResult?) = apply { this.expectedResult = expectedResult }

        fun build(): ValidationDetails {
            return ValidationDetails(
                testStepId, expectedResult ?: error("Expected result object must be defined in the validation details")
            )
        }
    }
}
